drivers/gpio/gpio_xlnx_axi.c

```diff
@@ -247,76 +242,111 @@ static const struct gpio_driver_api gpio_xlnx_axi_driver_api = {
-	IF_ENABLED(UTIL_AND(						\
-			DT_INST_PROP_OR(n, xlnx_is_dual, 1),		\
-			GPIO_XLNX_AXI_GPIO2_HAS_COMPAT_STATUS_OKAY(n)	\
-			),						\
-		(GPIO_XLNX_AXI_GPIO2_INIT(n)));
-
-#define GPIO_XLNX_AXI_GPIO2_INIT(n)					\
-	static struct gpio_xlnx_axi_data gpio_xlnx_axi_##n##_2_data = {	\
-		.dout = DT_INST_PROP_OR(n, xlnx_dout_default_2, 0),	\
-		.tri = DT_INST_PROP_OR(n, xlnx_tri_default_2,		\
-				       GENMASK(MAX_GPIOS - 1, 0)),	\
-	};								\
-									\
-	static const struct gpio_xlnx_axi_config			\
-		gpio_xlnx_axi_##n##_2_config = {			\
-		.common = {						\
-			.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_NGPIOS(\
-				DT_INST_PROP_OR(n, xlnx_gpio2_width,	\
-						MAX_GPIOS)),		\
-		},							\
-		.base = DT_INST_REG_ADDR(n) + GPIO2_DATA_OFFSET,	\
-		.all_inputs = DT_INST_PROP_OR(n, xlnx_all_inputs2, 0),	\
-		.all_outputs = DT_INST_PROP_OR(n, xlnx_all_outputs2, 0),\
-	};								\
-									\
-	DEVICE_DT_DEFINE(DT_CHILD(DT_DRV_INST(n), gpio2),		\
-			&gpio_xlnx_axi_init,				\
-			NULL,						\
-			&gpio_xlnx_axi_##n##_2_data,			\
-			&gpio_xlnx_axi_##n##_2_config,			\
-			PRE_KERNEL_1,					\
-			CONFIG_GPIO_INIT_PRIORITY,			\
-			&gpio_xlnx_axi_driver_api);
-
-#define GPIO_XLNX_AXI_INIT(n)						\
-	static struct gpio_xlnx_axi_data gpio_xlnx_axi_##n##_data = {	\
-		.dout = DT_INST_PROP_OR(n, xlnx_dout_default, 0),	\
-		.tri = DT_INST_PROP_OR(n, xlnx_tri_default,		\
-				       GENMASK(MAX_GPIOS - 1, 0)),	\
-	};								\
-									\
-	static const struct gpio_xlnx_axi_config			\
-		gpio_xlnx_axi_##n##_config = {				\
-		.common = {						\
-			.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_NGPIOS(\
-				DT_INST_PROP_OR(n, xlnx_gpio_width,	\
-						MAX_GPIOS)),		\
-		},							\
-		.base = DT_INST_REG_ADDR(n),				\
-		.all_inputs = DT_INST_PROP_OR(n, xlnx_all_inputs, 0),	\
-		.all_outputs = DT_INST_PROP_OR(n, xlnx_all_outputs, 0),	\
-	};								\
-									\
-	DEVICE_DT_INST_DEFINE(n,					\
-			&gpio_xlnx_axi_init,				\
-			NULL,						\
-			&gpio_xlnx_axi_##n##_data,			\
-			&gpio_xlnx_axi_##n##_config,			\
-			PRE_KERNEL_1,					\
-			CONFIG_GPIO_INIT_PRIORITY,			\
-			&gpio_xlnx_axi_driver_api);			\
+#define GPIO_XLNX_AXI_GPIO2_COND_INIT(n)				 \
+	IF_ENABLED(UTIL_AND(						 \
+			   DT_INST_PROP_OR(n, xlnx_is_dual, 1),		 \
+			   GPIO_XLNX_AXI_GPIO2_HAS_COMPAT_STATUS_OKAY(n) \
+			   ),						 \
+		   (GPIO_XLNX_AXI_GPIO2_INIT(n)));
+
+#ifdef CONFIG_MMU
+#define GPIO_XLNX_AXI_GPIO2_INIT(n)					      \
+	static struct gpio_xlnx_axi_data gpio_xlnx_axi_##n##_2_data = {	      \
+		.dout = DT_INST_PROP_OR(n, xlnx_dout_default_2, 0),	      \
+		.tri = DT_INST_PROP_OR(n, xlnx_tri_default_2,		      \
+				       GENMASK(MAX_GPIOS - 1, 0)),	      \
+	};								      \
+									      \
+	static const struct gpio_xlnx_axi_config			      \
+			    gpio_xlnx_axi_##n##_2_config = {		      \
+		._mmio = {						      \
+			.phys_addr = DT_INST_REG_ADDR(n) + GPIO2_DATA_OFFSET, \
+			.size = DT_INST_REG_SIZE			      \
+		},							      \
```

Comment:
This is clearly wrong. You are adding an offset, but using the size without taking the offset into consideration.